src/scipp/coords.py

```diff
@@ -2,247 +2,480 @@
+        outputs = self._to_dict(outputs)
+        return {
+            name: _Coord(dense=var, event=None, destination=_Destination.coord)
+            for name, var in outputs.items()
+        }
+
+    def _compute_with_events(self, inputs):
+        args = {
+            name: coord.event if coord.has_event else coord.dense
+            for name, coord in inputs.items()
+        }
+        outputs = self._to_dict(self._func(**args))
+        # Dense outputs may be produced as side effects of processing event
+        # coords.
+        outputs = {
+            name: _Coord(dense=var if var.bins is None else None,
+                         event=var if var.bins is not None else None,
+                         destination=_Destination.coord)
+            for name, var in outputs.items()
+        }
+        return outputs
+
+    def _without_unrequested(self, d: Dict[str, Any]) -> Dict[str, Any]:
+        return {key: val for key, val in d.items() if key in self.out_names}
+
+    def _to_dict(self, output) -> Dict[str, Variable]:
+        if not isinstance(output, dict):
+            if len(self.out_names) != 1:
+                raise TypeError('Function returned a single output but '
+                                f'{len(self.out_names)} were expected.')
+            return {self.out_names[0]: output}
+        return output
+
+    @property
+    def dependencies(self) -> Tuple[str]:
+        return self._arg_names
+
+    @property
+    def func_name(self) -> str:
+        return self._func.__name__
+
+    def __str__(self):
+        return f'Compute {self._format_out_names()} = {self._func.__name__}' \
+               f'({", ".join(self._arg_names)})'
+
+
+def _make_rule(products, producer) -> _Rule:
+    if isinstance(producer, str):
+        return _RenameRule(products, producer)
+    return _ComputeRule(products, producer)
+
+
+def _non_duplicate_rules(rules: Mapping[str, _Rule]) -> List[_Rule]:
+    already_used = set()
+    result = []
+    for rule in filter(lambda r: r not in already_used,
+                       map(lambda n: rules[n], _sort_topologically(rules))):
+        already_used.add(rule)
+        result.append(rule)
+    return result
+
+
+def _convert_to_rule_graph(graph: GraphDict) -> Dict[str, _Rule]:
+    rule_graph = {}
+    for products, producer in graph.items():
+        products = (products, ) if isinstance(products, str) else tuple(products)
+        rule = _make_rule(products, producer)
+        for product in products:
+            if product in rule_graph:
+                raise ValueError(
+                    f'Duplicate output name defined in conversion graph: {product}')
+            rule_graph[product] = rule
+    return rule_graph
+
+
+def _sort_topologically(graph: Mapping[str, _Rule]) -> Iterable[str]:
+    yield from TopologicalSorter(
+        {out: rule.dependencies
+         for out, rule in graph.items()}).static_order()
+
+
+def _is_meta_data(name: str, data: DataArray) -> bool:
```

Comment:
I often prefer to use `da` for data arrays, since `data` is also a property of data arrays.